spec/unit/typesafe/config/config_value_spec.rb

```diff
@@ -0,0 +1,503 @@
+  a_list = SimpleConfigList.new(TestUtils.fake_origin, a_values)
+
+  same_as_a_values = [1, 2, 3].map { |i| TestUtils.int_value(i) }
+  same_as_a_list = SimpleConfigList.new(TestUtils.fake_origin, same_as_a_values)
+
+  b_values = [4, 5, 6].map { |i| TestUtils.int_value(i) }
+  b_list = SimpleConfigList.new(TestUtils.fake_origin, b_values)
+
+  context "a_list equals a_list" do
+    let(:first_object) { a_list }
+    let(:second_object) { a_list }
+    include_examples "object_equality"
+  end
+
+  context "a_list equals same_as_a_list" do
+    let(:first_object) { a_list }
+    let(:second_object) { same_as_a_list }
+    include_examples "object_equality"
+  end
+
+  context "a_list doesn't equal b_list" do
+    let(:first_object) { a_list }
+    let(:second_object) { b_list }
+    include_examples "object_inequality"
+  end
+end
+
+describe "ConfigReference equality" do
+  a = TestUtils.subst("foo")
+  same_as_a = TestUtils.subst("foo")
+  b = TestUtils.subst("bar")
+  c = TestUtils.subst("foo", true)
+
+  specify "testing values are of the right type" do
+    expect(a).to be_instance_of(ConfigReference)
+    expect(b).to be_instance_of(ConfigReference)
+    expect(c).to be_instance_of(ConfigReference)
+  end
+
+  context "a equals a" do
+    let(:first_object) { a }
+    let(:second_object) { a }
+    include_examples "object_equality"
+  end
+
+  context "a equals same_as_a" do
+    let(:first_object) { a }
+    let(:second_object) { same_as_a }
+    include_examples "object_equality"
+  end
+
+  context "a doesn't equal b" do
+    let(:first_object) { a }
+    let(:second_object) { b }
+    include_examples "object_inequality"
+  end
+
+  context "a doesn't equal c, an optional substitution" do
+    let(:first_object) { a }
+    let(:second_object) { c }
+    include_examples "object_inequality"
+  end
+end
+
+describe "ConfigConcatenation equality" do
+  a = TestUtils.subst_in_string("foo")
+  same_as_a = TestUtils.subst_in_string("foo")
+  b = TestUtils.subst_in_string("bar")
+  c = TestUtils.subst_in_string("foo", true)
+
+  specify "testing values are of the right type" do
+    expect(a).to be_instance_of(ConfigConcatenation)
+    expect(b).to be_instance_of(ConfigConcatenation)
+    expect(c).to be_instance_of(ConfigConcatenation)
+  end
+
+  context "a equals a" do
+    let(:first_object) { a }
+    let(:second_object) { a }
+    include_examples "object_equality"
+  end
+
+  context "a equals same_as_a" do
+    let(:first_object) { a }
+    let(:second_object) { same_as_a }
+    include_examples "object_equality"
+  end
+
+  context "a doesn't equal b" do
+    let(:first_object) { a }
+    let(:second_object) { b }
+    include_examples "object_inequality"
+  end
+
+  context "a doesn't equal c, an optional substitution" do
+    let(:first_object) { a }
+    let(:second_object) { c }
+    include_examples "object_inequality"
+  end
+end
+
+describe "ConfigDeflayedMerge equality" do
```

Comment:
Not a huge deal and I don't think we should block the PR on it, but we might wanna change this before the PR gets merged.